### PR TITLE
Source convenience macros

### PR DESCRIPTION
**Pull request by broad-well**
## Features

* [x] `loadOpt` macro for loading from parameters to `JObject`
* [x] `assignFrom` macro for loading optional fields from `JObject` to Nim objects
* [ ] ~~`CommandHandlerTable` and related abstractions for command declarations~~
* [ ] ~~Outward DSL format similar to [Discordnim fork](https://github.com/broad-well/discordnim), except with more integrated features such as parameter validation and help text~~

### CODE REVIEW

**Comment by krisppurg** [comment]
I'm not quite too sure about the whole DSL thing, because I guess it seems out of scope of the library, since the library is like mainly for discord bots and some things are probably opinionated, since they are both a different kettle of fish.

I know it's for simplicity of course, but I think you should either maybe remove the DSL thing or you could make a separate repository for that.

Overall, I will be testing the code.

**Comment by broad-well** [comment]
Okay. I will most likely split that into my own library.

In terms of conformity, I tried to replicate the original behavior to the greatest extent possible, and the macro's expansions are almost 100% identical to what was written before. I took note of some discrepancies, and I will comment on the specific lines to which they apply.

File: dimscord/restapi.nim [hunk 1]
```diff
@@ -347,65 +347,41 @@ proc getChannelPins*(api: RestApi, channel_id: string): Future[seq[Message]] {.a
...
-    if bitrate.isSome:
-        payload["bitrate"] = %get(bitrate)
-    if user_limit.isSome:
-        payload["user_limit"] = %get(user_limit)
-    if permission_overwrites.isSome:
-        var perms: seq[JsonNode] = @[]
-
-        for ow in get(permission_overwrites):
-            perms.add(%ow)
-        payload["permission_overwrites"] = %perms
```
Comment (broad-well) [inline dimscord/restapi.nim:375]: From my testing, it seems that `%` serializes the elements of a `seq` recursively, so this is not necessary. The macro replacement is `payload["permission_overwrites"] = %get(permission_overwrites)`.

File: dimscord/restapi.nim [hunk 2]
```diff
@@ -439,54 +415,28 @@ proc deleteGuild*(api: RestApi, guild_id: string): Future[void] {.async.} =
...
-        afk_chan_id: Option[string] = none(string);
-        afk_timeout: Option[int] = none(int);
-        icon: Option[string] = none(string);
-        owner_id: Option[string] = none(string);
-        splash: Option[string] = none(string);
-        banner: Option[string] = none(string);
-        system_chan_id: Option[string] = none(string);
-        rules_chan_id: Option[string] = none(string);
-        public_update_chan_id: Option[string] = none(string);
-        prefered_locale: Option[string] = none(string); reason: string = ""): Future[Guild] {.async.} =
```
Comment (broad-well) [inline dimscord/restapi.nim:456]: The official specs accept `preferred_locale`. This has been fixed.

File: dimscord/objects.nim [hunk 3]
```diff
@@ -577,74 +618,55 @@ proc newEmoji*(data: JsonNode): Emoji =
...
     if data.hasKey("party"):
-        result.party = some((id: "", size: ""))
-        if data["party"].hasKey("id"):
-            get(result.party).id = data["party"]["id"].str
-        if data["party"].hasKey("size"):
-            # TODO: Parse MaxSize too?
-            get(result.party).size = $data["party"]["size"][0].getInt()
+        result.party = some((
+            id: data["party"]{"id"}.getStr,
+            size: data["party"]{"size"}.getStr
```
Comment (broad-well) [inline dimscord/objects.nim:653]: It seems that the `getInt` line change occurred on your `devel` branch. This needs to be reflected in my changes.

File: dimscord/objects.nim [hunk 4]
```diff
@@ -577,74 +618,55 @@ proc newEmoji*(data: JsonNode): Emoji =
...
+            large_image <- it.str
 
     if data.hasKey("secrets"):
-        result.secrets = some((join: "", spectate: "", match: ""))
-        if data["secrets"].hasKey("join"):
-            get(result.secrets).join = data["secrets"].str
-        if data["secrets"].hasKey("spectate"):
-            get(result.secrets).spectate = data["spectate"].str
-        if data["secrets"].hasKey("match"):
-            get(result.secrets).match = data["match"].str
```
Comment (broad-well) [inline dimscord/objects.nim:642]: Seemed like there were typos here? `data["secrets"]["join"]`, `data["secrets"]["spectate"]`, and `data["secrets"]["match"]` match up with the predicates. These are assumed in my changes.

File: dimscord/restapi.nim [hunk 5]
```diff
@@ -523,21 +470,15 @@ proc editGuildRole*(api: RestApi, guild_id: string, role_id: string;
...
+    let payload = %*{}
+
+    payload.loadOpt(name, color, hoist, mentionable)
     if pobj.isSome:
         payload["permissions"] = %(+(get(pobj)))
 
     result = (await api.request("POST", endpointGuildRoles(guild_id), $(payload))).newRole
 
-proc editGuildRolePosition*(api: RestApi, guild_id: string, role_id: string; position: Option[int] = none(int); reason: string = ""): Future[seq[Role]] {.async.} =
+proc editGuildRolePosition*(api: RestApi, guild_id: string, role_id: string, position: int, reason: string = ""): Future[seq[Role]] {.async.} =
```
Comment (krisppurg) [inline dimscord/restapi.nim:481]: The optional position is needed. (see: https://discordapp.com/developers/docs/resources/guild#modify-guild-role-positions)

File: dimscord/restapi.nim [hunk 6]
```diff
@@ -554,39 +495,29 @@ proc getGuildInvites*(api: RestApi, guild_id: string): Future[seq[InviteMetadata
...
-        payload["nick"] = %nick
-    if roles.isSome:
-        payload["roles"] = %roles
-    if mute.isSome:
-        payload["mute"] = %mute
-    if deaf.isSome:
-        payload["deaf"] = %deaf
-    if channel_id.isSome:
-        payload["channel_id"] = %channel_id
+    var payload = %*{}
```
Comment (krisppurg) [inline dimscord/restapi.nim:512]: I'm not sure, about this, but I believe that `%*{}` should be `newJObject()`.

File: dimscord/restapi.nim [hunk 7]
```diff
@@ -637,7 +568,7 @@ proc getGuildMember*(api: RestApi, guild_id: string, user_id: string): Future[Me
 proc getGuildMembers*(api: RestApi, guild_id: string, limit: int = 1, after: string = "0"): Future[seq[Member]] {.async.} =
     ## Gets a list of a guild's members.
     result = @[]
-    let mems = (await api.request("GET", endpointGuildMembers(guild_id)))
+    let mems = (await api.request("GET", endpointGuildChannels(guild_id)))
```
Comment (krisppurg) [inline dimscord/restapi.nim:571]: It should not be `endpointGuildChannels`.

Comment (krisppurg) [inline dimscord/restapi.nim:571]: It should not be `endpointGuildChannels`.

**Comment by broad-well** [comment]
See `230c5bf` for fixes

File: dimscord/objects.nim (hunk 3 above)
Comment (broad-well) [reply dimscord/objects.nim:653]: Although the type of `size` is still `string`, as defined in objects.nim:174?
```nim
party*: Option[tuple[id: string, size: string]]
```

**Comment by krisppurg** [comment]
Didn't know that, but thats cool.

**Comment by krisppurg** [comment]
There is more bugs with the objects.nim file, first when you either receive `GUILD_CREATE`, there is a chance that when calling assignFrom the program just crashes with no error at all.

**Comment by broad-well** [comment]
> There is more bugs with the objects.nim file, first when you either receive `GUILD_CREATE`, there is a chance that when calling assignFrom in `newGuild` the program just crashes with no error at all.

I ran `expandMacros` on the `assignFrom` invocation in `newGuild`--here's the expanded syntax:

```nim
if hasKey(data, "afk_timeout"):
  result.afk_timeout = some(getInt(data["afk_timeout"], 0))
if hasKey(data, "permissions"):
  result.permissions = some(getInt(data["permissions"], 0))
if hasKey(data, "widget_channel_id"):
  result.widget_channel_id = some(data["widget_channel_id"].str)
if hasKey(data, "joined_at"):
  result.joined_at = some(data["joined_at"].str)
if hasKey(data, "large"):
  result.large = some(data["large"].bval)
if hasKey(data, "unavailable"):
  result.unavailable = some(data["unavailable"].bval)
if hasKey(data, "member_count"):
  result.member_count = some(getInt(data["member_count"], 0))
if hasKey(data, "premium_subscription_count"):
  result.premium_subscription_count = some(
      getInt(data["premium_subscription_count"], 0))
if hasKey(data, "max_presences") and
  not (data["max_presences"].kind == JNull):
  result.max_presences = some(getInt(data["max_presences"], 0))
if hasKey(data, "embed_channel_id") and
  not (data["embed_channel_id"].kind == JNull):
  result.embed_channel_id = data["embed_channel_id"].str
if hasKey(data, "icon") and
  not (data["icon"].kind == JNull):
  result.icon = some(data["icon"].str)
if hasKey(data, "splash") and
  not (data["splash"].kind == JNull):
  result.splash = some(data["splash"].str)
if hasKey(data, "afk_channel_id") and
  not (data["afk_channel_id"].kind == JNull):
  result.afk_channel_id = some(data["afk_channel_id"].str)
if hasKey(data, "application_id") and
  not (data["application_id"].kind == JNull):
  result.application_id = some(data["application_id"].str)
if hasKey(data, "system_channel_id") and
  not (data["system_channel_id"].kind == JNull):
  result.system_channel_id = some(data["system_channel_id"].str)
if hasKey(data, "vanity_url_code") and
  not (data["vanity_url_code"].kind == JNull):
  result.vanity_url_code = some(data["vanity_url_code"].str)
if hasKey(data, "description") and
  not (data["description"].kind == JNull):
  result.description = some(data["description"].str)
if hasKey(data, "banner") and
  not (data["banner"].kind == JNull):
  result.banner = some(data["banner"].str)
if hasKey(data, "discovery_splash") and
  not (data["discovery_splash"].kind == JNull):
  result.discovery_splash = some(data["discovery_splash"].str)
```
The only difference from your original code seems to be that it checks for certain required fields, such as `description` and `banner`. Could the crash be associated with the use of `str` rather than `getStr`?

**Comment by krisppurg** [comment]
I think so, I'll test it.

**Comment by krisppurg** [comment]
I have tested the code and it works just fine.

**Comment by krisppurg** [comment]
@broad-well ?

**Comment by krisppurg** [comment]
(This issue hasn't been known yet why it has been crashing like that, so since it may take more time, I won't be able to push changes.)

**Pull request closed without merging**